### PR TITLE
Telemetry interface and plan cleanups

### PR DESCRIPTION
**Pull request by kmdalal**
### PLEXIL plan changes

- Every call to `PanTiltMoveJoints` that was essentially just a pan was replaced with `Pan`.
- `ActionGoalStateDemo` was fixed and improved.  Created macros for the action states.
- `ReferenceMission2` was slightly improved, and tested.
- Many (stubbed) Lookups specific to `EuropaMission.plp` were replaced with (stubbed) literal values.
- Calls to `set_checkpoint` in `EuropaMission.plp` were fixed.
- Calls to the obsolete `dig_linear` command were replaced with `scoop_linear`.
- `StowSequence` was renamed to `SafeStow`.
- Calls to `ArmUnstow` followed by `ArmStow` were replaced with `SafeStow`, or in some cases just `ArmStow`.
- Duplicate Lookup and `LibraryAction` declarations were removed.
- Improved plan formatting, comments, and ROS messages.  Renamed some identifiers.
- The old plan header `plexil_defs.h` was renamed to `europa_defs.h`, with most of its contents eliminated or moved to `plan-interface.h`.
- PLEXIL's full checkpointing interface was added to `common-interface.h`, with comments.
- The macros `BASE_LINK` and `SCOOP_TIP` were replaced with the newer `BASE_FRAME` and `TOOL_FRAME`, respectively.

### PLEXIL adapter changes

- The header `joint_support.h` was obsoleted and removed.
- The `STATE_STUB` C macro was eliminated.
- Plans are compiled into formatted XML (the `-p` option to `plexilc`).
- Installation of the OceanWATERS config file was removed from the OWLAT build. (It was harmless anyway).

### Reviewer suggestions
- @AstroStucky : code review, OWLAT build, OceanWATERS build and regression tests.
- @Keeganfn : code review, OceanWATERS build and revised plan tests.
- Both reviewers feel free to test anything!
- Don't be alarmed by the number of files! Most have only trivial changes such as a renamed identifier or header file.

### OceanWATERS build and tests
- Have all other repos on the `master` branch.  Then `catkin clean` followed by `catkin build`.  
- Refer to [Release 11.1 tests](https://babelfish.arc.nasa.gov/confluence/pages/viewpage.action?pageId=220561639) for the tests below.
- Recommend maximizing Gazebo realtime rate to speed up testing: `gz physics --update-rate=1200`
- Test revised plans:
   - Launch the Atacama world and plan executive and test `ActionGoalStateDemo` : test 10.10.
   - In the same launches, test the `EuropaMission` plan: test 10.11.
   - Relaunch both and test `ReferenceMission2`: test 10.14.
- Regression tests:
   - Restart the Atacama world and test executive and run `TestTelemetry`: test 10.5.
   - In the same launches, run `ReferenceMission1`: test 10.13 (you don't have to use the plan selection GUI).

### OWLAT build
- Have all other repos on the `master` branch.  Then `catkin clean` followed by `catkin build -DOWLAT=ON`.  (Note: if you've done a clean build of OceanWATERS, only `ow_plexil` needs to be rebuilt clean here).
- A test seems unnecessary since the only change made to OWLAT plans was renaming `StowSequence` to `SafeStow`.

### CODE REVIEW

**Comment by Keeganfn** [comment]
In ```ReferenceMission2.plp``` if an Antenna Fault is injected during ```ImageLandingSite()``` then the plan will freeze and not move forward even after the fault is cleared. I believe this is tied to the SynchronousCommand call in ```PanTiltMoveJoints.plp```.

This is probably outside the scope for this PR, but might be worth noting at the top of ```ReferenceMission2.plp```.

File: ow_plexil/src/plans/plan-interface.h [hunk 1]
```diff
@@ -26,6 +22,20 @@
...
+// Action goal statuses
+
+#define ACTION_PENDING 0
+#define ACTION_ACTIVE 1
+#define ACTION_PREEMPTED 2
+#define ACTION_SUCCEEDED 3
+#define ACTION_ABORTED 4
+#define ACTION_REJECTED 5
+#define ACTION_PREEMPTING 6
+#define ACTION_RECALLING 7
```
Comment (Keeganfn) [inline ow_plexil/src/plans/plan-interface.h:36]: I think Thomas mentioned that  ACTION_RECALLING, ACTION_RECALLED and ACTION_LOST are not currently supported in OW. It may be a good idea to mention this in a comment.

**Review by Keeganfn — COMMENTED** [review]
The changes all look great, and many of the files are much easier to read now! 

Review Summary:
- Minor comment suggestions
- Found an issue with the Antenna Fault freezing PLEXIL in ReferenceMission2 that is most likely unrelated to this PR.
- Similar Antenna Fault issue in AntennaGoalStateDemo.

~Confluence is down for maintenance until Monday, but I ran the test plans and tested all the edge cases I could think of. I will review the official tests early Monday morning to make sure I didn't miss anything important before approving the review!~

Finished review, it overall looks great! The largest concern I have is the Antenna Fault issues in ReferenceMission2 and AntennaGoalStateDemo. It would be useful to see if Thomas or you can recreate those. For me it happened every time I ran the plans.

It is possible that those issues are outside the scope of this PR or could be PLEXIL related.

**Comment by Keeganfn** [comment]
In ```ActionGoalStateDemo``` I see a similar issue to ```ReferenceMission2```. When an AntennaPan fault is injected, the antenna stops moving and crashes the PLEXIL node. 

To fix this, I tried using a structure similar to ```FaultHandlingPattern1``` which did fix the antenna permanently stopping from a fault, but the arm did not mind its start condition for AntennaPan fault and continued execution.

Im not sure if this is a PLEXIL issue or something else. It seems strange to me that only the Antenna commands and lookups seem to be causing issues.

Are you able to recreate this behavior?

**Comment by AstroStucky** [comment]
@kmdalal does testing this PR serve as the final release test for 11.1? Should we go ahead and update the Confluence page with test results, or hold of for a later test repeat?

**Comment by kmdalal** [comment]
> @kmdalal does testing this PR serve as the final release test for 11.1? Should we go ahead and update the Confluence page with test results, or hold of for a later test repeat?

This is not a test for Release 11.1, as there is more to come for that release that would entail a retest of these features.

**Comment by AstroStucky** [comment]
Releast Test 10.6 - The "Enter any additional plan to be run (or use the GUI):" prompt did not show following TestTelemetry.plx from 10.5 and it will not take in input. Could TestTelemetry.plx be blocking?

**Comment by kmdalal** [comment]
> In `ReferenceMission2.plp` if an Antenna Fault is injected during `ImageLandingSite()` then the plan will freeze and not move forward even after the fault is cleared. I believe this is tied to the SynchronousCommand call in `PanTiltMoveJoints.plp`.
> 
> This is probably outside the scope for this PR, but might be worth noting at the top of `ReferenceMission2.plp`.

Good to know this, but yes it's out of scope for this PR.  The 10.14 test has a note that fault injection has not yet been added to the test.  Your intuition of the cause of the problem is probably correct, and also the plan might be held up because the calling node is in a _checked_ Sequence (the default).  I'm creating a new Jira to debug this plan.

**Comment by kmdalal** [comment]
> Releast Test 10.6 - The "Enter any additional plan to be run (or use the GUI):" prompt did not show following **TestTelemetry.plx** from 10.5 and it will not take in input. Could **TestTelemetry.plx** be blocking?

That's a simple plan that should complete in a second.  I've run it consecutively without problems, both from the prompt and GUI.  Do you still have the output?

**Comment by Keeganfn** [comment]
> Releast Test 10.6 - The "Enter any additional plan to be run (or use the GUI):" prompt did not show following **TestTelemetry.plx** from 10.5 and it will not take in input. Could **TestTelemetry.plx** be blocking?

I thought this at first as well. The "Enter any additional plan to be run (or use the GUI):" print statement comes out in the wrong order for this plan because it executes so fast. I was still able to submit additional plans through the GUI and the terminal (although the formatting was messed up).

**Comment by AstroStucky** [comment]
> That's a simple plan that should complete in a second. I've run it consecutively without problems, both from the prompt and GUI. Do you still have the output?

```
$ roslaunch ow_plexil ow_exec.launch plan:=TestTelemetry.plx
... logging to /usr/local/home/tstucky/.ros/log/3afa1eca-24cd-11ee-ae89-47c5bc6c583f/roslaunch-flynn.ndc.nasa.gov-66533.log
Checking log directory for disk usage. This may take a while.
Press Ctrl-C to interrupt
Done checking log file disk usage. Usage is <1GB.

started roslaunch server http://flynn.ndc.nasa.gov:34607/

SUMMARY
========

PARAMETERS
 * /owlat_flag: False
 * /rosdistro: noetic
 * /rosversion: 1.16.0

NODES
  /
    identify_sample_location (ow_plexil/identify_sample_location.py)
    ow_exec_node (ow_plexil/ow_exec_node)
    terminal_selection_node (ow_plexil/terminal_selection_node)

ROS_MASTER_URI=http://localhost:11311

process[ow_exec_node-1]: started with pid [66566]
process[terminal_selection_node-2]: started with pid [66567]
process[identify_sample_location-3]: started with pid [66568]
[ INFO:/ow_exec_node] Starting PLEXIL executive node...
[ INFO:/ow_exec_node] Executive node started, ready for PLEXIL plans.
[ INFO:/ow_exec_node] PLEXIL: Starting telemetry tests...
[ INFO:/ow_exec_node] PLEXIL: OceanWATERS is in use.
[ INFO:/ow_exec_node] PLEXIL: OWLAT is NOT in use.
[ INFO:/ow_exec_node] PLEXIL: Battery state of charge: 0.916168332099915
[ INFO:/ow_exec_node] PLEXIL: Battery remaining useful life: 2736
[ INFO:/ow_exec_node] PLEXIL: Battery temperature: 20.00427030471
[ INFO:/ow_exec_node] PLEXIL: Pan degees: -0.000402833981244534
[ INFO:/ow_exec_node] PLEXIL: Pan radians: -7.03077931163421e-06
[ INFO:/ow_exec_node] PLEXIL: Tilt degees: -1.0774056724989
[ INFO:/ow_exec_node] PLEXIL: Tilt radians: -0.0188042763647696
[ INFO:/ow_exec_node] PLEXIL: Position of arm joint 0 = 2.88384396887812
[ INFO:/ow_exec_node] PLEXIL: Velocity of arm joint 0 = -1.37985840233916e-17
[ INFO:/ow_exec_node] PLEXIL: Torque of arm joint 0 = 1.94032742110546
[ INFO:/ow_exec_node] PLEXIL: Acceleration of arm joint 0 = -1.7662207869633e-15
[ INFO:/ow_exec_node] PLEXIL: Position of arm joint 1 = -0.118486674196351
[ INFO:/ow_exec_node] PLEXIL: Velocity of arm joint 1 = -0.00101401819616559
[ INFO:/ow_exec_node] PLEXIL: Torque of arm joint 1 = 0
[ INFO:/ow_exec_node] PLEXIL: Acceleration of arm joint 1 = 3.96382576944463e-05
[ INFO:/ow_exec_node] PLEXIL: Position of arm joint 2 = -2.30926389122033e-14
[ INFO:/ow_exec_node] PLEXIL: Velocity of arm joint 2 = 2.73610935950927e-18
[ INFO:/ow_exec_node] PLEXIL: Torque of arm joint 2 = 2.30983112270337e-11
[ INFO:/ow_exec_node] PLEXIL: Acceleration of arm joint 2 = 1.26249774819644e-16
[ INFO:/ow_exec_node] PLEXIL: Position of arm joint 3 = 3.58796751365462
[ INFO:/ow_exec_node] PLEXIL: Velocity of arm joint 3 = 2.08516979214123e-17
[ INFO:/ow_exec_node] PLEXIL: Torque of arm joint 3 = 5.18060641684758
[ INFO:/ow_exec_node] PLEXIL: Acceleration of arm joint 3 = -5.05190426154258e-17
[ INFO:/ow_exec_node] PLEXIL: Position of arm joint 4 = 0
[ INFO:/ow_exec_node] PLEXIL: Velocity of arm joint 4 = 6.2007836861236e-20
[ INFO:/ow_exec_node] PLEXIL: Torque of arm joint 4 = 8.9237781439381e-14
[ INFO:/ow_exec_node] PLEXIL: Acceleration of arm joint 4 = -3.05540712471297e-16
[ INFO:/ow_exec_node] PLEXIL: Position of arm joint 5 = 1.54838456522445
[ INFO:/ow_exec_node] PLEXIL: Velocity of arm joint 5 = -4.69571586436097e-17
[ INFO:/ow_exec_node] PLEXIL: Torque of arm joint 5 = 3.20780631518987
[ INFO:/ow_exec_node] PLEXIL: Acceleration of arm joint 5 = 1.3833785025231e-14
[ INFO:/ow_exec_node] PLEXIL: Position of arm joint 6 = -1.49999999999678
[ INFO:/ow_exec_node] PLEXIL: Velocity of arm joint 6 = -1.42595147033051e-17
[ INFO:/ow_exec_node] PLEXIL: Torque of arm joint 6 = 6.28072538378947e-12
[ INFO:/ow_exec_node] PLEXIL: Acceleration of arm joint 6 = 4.71227972185752e-16
[ INFO:/ow_exec_node] PLEXIL: Arm end effector force/torque: {
[ INFO:/ow_exec_node] PLEXIL:   -2.64283911237321
[ INFO:/ow_exec_node] PLEXIL:   -15.0424811740287
[ INFO:/ow_exec_node] PLEXIL:   0.108325365044248
[ INFO:/ow_exec_node] PLEXIL:   0.103084114611153
[ INFO:/ow_exec_node] PLEXIL:   -0.0170862229548323
[ INFO:/ow_exec_node] PLEXIL:   -2.17004426531117
[ INFO:/ow_exec_node] PLEXIL: }
[ INFO:/ow_exec_node] PLEXIL: Arm pose: {
[ INFO:/ow_exec_node] PLEXIL:   0.689081129353541
[ INFO:/ow_exec_node] PLEXIL:   -0.52242467342376
[ INFO:/ow_exec_node] PLEXIL:   0.514781035298368
[ INFO:/ow_exec_node] PLEXIL:   -0.0228590034311636
[ INFO:/ow_exec_node] PLEXIL:   0.645560831871841
[ INFO:/ow_exec_node] PLEXIL:   -0.762888536527328
[ INFO:/ow_exec_node] PLEXIL:   0.0270140546801868
[ INFO:/ow_exec_node] PLEXIL: }
[ INFO:/ow_exec_node] PLEXIL: Telemetry tests finished.
TestLanderLights.plx
```

What you see at the very end is my attempt to input the next plan, despite the prompt not showing. **TestLanderLights.plx** did not start. I did not try sending a new plan via the GUI.

**Comment by kmdalal** [comment]
> In `ActionGoalStateDemo` I see a similar issue to `ReferenceMission2`. When an AntennaPan fault is injected, the antenna stops moving and crashes the PLEXIL node.

After many tries I was able to reproduce the frozen antenna but not the node crash.  I think I know why the antenna is freezing and am working on a  solution.

**Comment by kmdalal** [comment]
> What you see at the very end is my attempt to input the next plan, despite the prompt not showing. **TestLanderLights.plx** did not start. I did not try sending a new plan via the GUI.

I think I've seen the prompt not return if the last plan was sent from the GUI.  Do you recall?  Please try the plan again from the GUI, and if there's no prompt, try it (or another plan) from the GUI, and let me know if that worked.  In any case, this problem is out of scope of this PR.

**Comment by AstroStucky** [comment]
> > What you see at the very end is my attempt to input the next plan, despite the prompt not showing. **TestLanderLights.plx** did not start. I did not try sending a new plan via the GUI.
> 
> I think I've seen the prompt not return if the last plan was sent from the GUI. Do you recall? Please try the plan again from the GUI, and if there's no prompt, try it (or another plan) from the GUI, and let me know if that worked. In any case, this problem is out of scope of this PR.

**Comment by AstroStucky** [comment]
I seem to have closed the PR instead of cancelling my comment, apologies. This should reopen it.

**Comment by AstroStucky** [comment]
> The log I shared was calling the plan from the command line, and you can see that on the first line of the log.
> 
> I did go ahead and try from the GUI with **TestTelemtry.plx** as you suggested, and it resulted in no prompt appearing after plan completion. I then tried **TestLanderLight.plx** from the GUI, and again there is no prompt after that plan completed.

I investigated this further...
**A)** In the case of GUI called plans not resulting in a prompt, I can actually still input the next plan despite the lack of a prompt, and it runs fine. After typing in the name of a plan, I press enter, and the prompt will then show up after the input quickly followed by the plan's output. 

**B)** In the case of command line called plans resulting in no prompt. I cannot input the next plan, the console simply does not seem to read whatever I give it. 

These appears to be 2 different bugs. The form **A** looks to me like the prompt is not being flushed when it should be, but this is just a guess. I don't have a suggestion for what's happening with form **B** without looking at the code. 

I cannot say for sure whether these occurred before this PR, so I will defer to the author's judgement as to whether it is out of scope or not.

**Comment by AstroStucky** [comment]
## 10.8 Results
I did not see the soft limit warning or hard limit error. Reattempted 3 times with the same output resulting.
```
Enter any additional plan to be run (or use the GUI): 
TorqueTest.plx
[ INFO:/ow_exec_node] PLEXIL: Beginning over-torque test...
[ INFO:/ow_exec_node] ArmUnstow () started...
[ INFO:/ow_exec_node] ArmUnstow finished in state SUCCEEDED
[ INFO:/ow_exec_node] Starting GuardedMove(x=2.00, y=0.00, z=0.30, dir_x=0.00, dir_y=0.00,dir_z=1.00, search_dist=0.50)
[ INFO:/ow_exec_node] GuardedMove () started...
[ INFO:/ow_exec_node] GuardedMove finished in state SUCCEEDED
[ INFO:/ow_exec_node] ArmUnstow () started...
[ INFO:/ow_exec_node] ArmUnstow finished in state SUCCEEDED
[ INFO:/ow_exec_node] ArmStow () started...
[ INFO:/ow_exec_node] ArmStow finished in state SUCCEEDED
[ INFO:/ow_exec_node] PLEXIL: Over-torque test finished.
```

## 10.10 Results
Calling **ActionGoalStateDemo.plx** causing the arm to continually stow and unstow, but the antenna mast remains still. The entire time the plan runs, the following is printed in the ow_plexil log very rapidly. 
```
[ WARN:/ow_exec_node] Requested pan -345.000000 out of valid range [-183.345993 183.345993], rejecting PLEXIL command.
```
Since the antenna does not move, I could not properly perform the remainder of this step.

### 10.10 Redo
I retried the test in isolation from the other steps in the release test with a fresh simulation start. This time I did not see the warning, and the antenna did move. 

I think this is happening because the outcome of the plan depends on the current position of the antenna. Maybe that plan needs some initial phase where it sends both the arm and antenna to known positions before it begins its looping phase?

#### 10.10 Redo Results
My outcome did not match the expected for the two antenna faults, but the other faults made the plan behaved as described. 

##### ant_tilt_joint_locked_failure
I first ticked **ant_tilt_joint_locked_failure** to on, and it did stop the antenna. I ticked it off and it resumed moving. The second time I ticked it on, the antenna did not stop as described. I could not replicate this with follow-up runs of the plan, and it typically behaved as expected.

##### ant_pan_joint_locked_failure
If you toggle **ant_pan_joint_locked_failure** multiple times, you can make the antenna stop entirely regardless of the state of either antenna fault. When this occurs I have to restart the plan to get the antenna moving again.

## 10.14
Passed, and I did not see either of the two warnings mentioned in the outcome following the start of **TaskDiscardSample**.

**Comment by kmdalal** [comment]
@AstroStucky, test 10.8 is not part of this PR, and the behavior you describe is a normal outcome.

For 10.10, the out of range error is a symptom of the old version of the plan -- not sure how you could have gotten it.  The tilt fault behavior is mysterious.  @Keeganfn reported the pan fault error.  I'm investigating all of these.

**Comment by AstroStucky** [comment]
Whoops. I misunderstood. Looks like I did some extra testing then.

File: ow_plexil/src/plexil-adapter/OwAdapter.cpp [hunk 2]
```diff
@@ -32,61 +32,32 @@ using std::unique_ptr;
...
 
+  if (state_name == "TrenchIdentified") {
+    value_out = true;
+  }
+  else if (state_name == "ExcavationTimeout") {
+    value_out = 10;
+  }
+  else if (state_name == "CollectAndTransferTimeout") {
+    value_out = 10;
+  }
```
Comment (AstroStucky) [inline ow_plexil/src/plexil-adapter/OwAdapter.cpp:51]: Should this generate output saying this value is stubbed and therefore is simply a constant?

**Review by AstroStucky — COMMENTED** [review]
One suggestion about angle wrapping. My other comments are simply asking for clarification. 

All tests pass except 10.10. Please let me know if you could use further testing on that or any other steps.

File: ow_plexil/src/plans/plan-interface.h (hunk 1 above)
Comment (kmdalal) [reply ow_plexil/src/plans/plan-interface.h:36]: Done, will be in next commit.

File: ow_plexil/src/plexil-adapter/OwAdapter.cpp (hunk 2 above)
Comment (kmdalal) [reply ow_plexil/src/plexil-adapter/OwAdapter.cpp:51]: Output could be repeated verbosely, and is not that useful, since it supports only one notional illustrative plan that is documented to be highly stubbed.

File: ow_plexil/src/plans/EuropaMission.plp [hunk 3]
```diff
@@ -8,8 +8,12 @@
 // support for crash recovery and plan persistence using PLEXIL's checkpointing
 // facility.
 
-#include "plexil_defs.h"
 #include "plan-interface.h"
+#include "europa_defs.h"
+
+// Maximum number of crashes before assuming something has gone very wrong and
+// attempting to offload all data
+#define MAX_CRASHES 10
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plans/EuropaMission.plp:16]: Crashing of the executive.

**Comment by kmdalal** [comment]
@AstroStucky, thanks for investigating the missing plan prompt problem.  It's out of this PR's scope, so I'm creating  a separate Jira for it.  It's a minor issue that I don't think needs to be fixed for 11.1.

File: ow_plexil/src/plans/ActionGoalStateDemo.plp [hunk 4]
```diff
@@ -1,47 +1,52 @@
...
+      endif;
 
-    Pattern2UnstowStow:
-    {
-      RepeatCondition true;
-      StartCondition !Lookup(ArmFault);
-      InvariantCondition !Lookup(ArmFault);
+      if (NewAngle > 180) {
+        NewAngle = NewAngle - 360;
+      }
```
[the remark this answers is not in the document]
Comment (kmdalal) [reply ow_plexil/src/plans/ActionGoalStateDemo.plp:51]: I tried this suggestion, and for some reason it left the antenna stationary, i.e. the computed pan was always the same number.  I'm deferring work on this plan to a new Jira.

**Comment by kmdalal** [comment]
@AstroStucky and @Keeganfn : I've created Jiras for the reported problems that are out of scope here:

- OW-1188: ActionGoalStateDemo issues
- OW-1189: Plan prompt not always returning
- OW-1190: ReferenceMission2 issues

Let me know if you have any more comments or requests.

**Review by Keeganfn — APPROVED** [review]

File: ow_plexil/src/plans/ActionGoalStateDemo.plp (hunk 4 above)
[the remark this answers is not in the document]
Comment (AstroStucky) [reply ow_plexil/src/plans/ActionGoalStateDemo.plp:51]: It definitely changes the logic flow, so there could be something happening I did not anticipate. Seeing as the test has other issues to resolved later on, I am perfectly fine with that.

**Review by AstroStucky — APPROVED** [review]
New changes were very minor, so I skipped a retest. My comments have all been addressed to my satisfaction. Great work!
